rm/rm-server/src/test/java/org/ow2/proactive/junitUtils/TimedOutTestsListener.java

```diff
@@ -0,0 +1,186 @@
+import java.io.StringWriter;
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Map;
+import org.junit.runner.notification.Failure;
+import org.junit.runner.notification.RunListener;
+
+/**
+ * JUnit run listener which prints full thread dump into System.err
+ * in case a test is failed due to timeout.
+ * @author Xiaoyan Lin (Netty project)
+ */
+public class TimedOutTestsListener extends RunListener {
+
+    static final String TEST_TIMED_OUT_PREFIX = "test timed out after";
+
+    private static String INDENT = "    ";
+
+    private final PrintWriter output;
+
+    public TimedOutTestsListener() {
+        this.output = new PrintWriter(System.err);
+    }
+
+    public TimedOutTestsListener(PrintWriter output) {
+        this.output = output;
+    }
+
+    @Override
+    public void testFailure(Failure failure) throws Exception {
+        if (failure != null && failure.getMessage() != null
+                && failure.getMessage().startsWith(TEST_TIMED_OUT_PREFIX)) {
+            output.println("====> TEST TIMED OUT. PRINTING THREAD DUMP. <====");
+            output.println();
+            output.print(buildThreadDiagnosticString());
+            output.flush();
+        }
+    }
+
+    public static String buildThreadDiagnosticString() {
+        StringWriter sw = new StringWriter();
+        PrintWriter output = new PrintWriter(sw);
+
+        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss,SSS");
+        output.println(String.format("Timestamp: %s", dateFormat.format(new Date())));
+        output.println();
+        output.println(buildThreadDump());
+
+        String deadlocksInfo = buildDeadlockInfo();
+        if (deadlocksInfo != null) {
+            output.println("====> DEADLOCKS DETECTED <====");
+            output.println();
+            output.println(deadlocksInfo);
+        }
+
+        return sw.toString();
+    }
+
+    static String buildThreadDump() {
+        StringBuilder dump = new StringBuilder();
+        Map<Thread, StackTraceElement[]> stackTraces = Thread.getAllStackTraces();
+        for (Map.Entry<Thread, StackTraceElement[]> e : stackTraces.entrySet()) {
+            Thread thread = e.getKey();
+            dump.append(String.format(
+                    "\"%s\" %s prio=%d tid=%d %s\njava.lang.Thread.State: %s",
+                    thread.getName(),
+                    thread.isDaemon() ? "daemon" : "",
+                    thread.getPriority(),
+                    thread.getId(),
+                    Thread.State.WAITING.equals(thread.getState()) ?
+                            "in Object.wait()" :
+                            (thread.getState().name() == null? "" : thread.getState().name().toLowerCase()),
```

Comment:
I think that the complexity of this method is too high which make it hardly readable and harder to debug. Maybe you could create a private method for the string formatting ?